tests/testsuite/cargo_command.rs

```diff
@@ -341,6 +341,58 @@ fn cargo_subcommand_env() {
         .run();
 }
 
+#[cfg(unix)]
+#[cargo_test]
+fn cargo_cmd_bin_path_env_interaction() {
+    use std::io::Write;
+    use std::os::unix::fs::PermissionsExt;
+
+    let root = paths::root();
+    let bin = root.join("bin");
+    fs::create_dir(&bin).unwrap();
+    fs::File::create(&bin.join("cargo-foo"))
+        .unwrap()
+        .write_all(
+            br#"!/usr/bin/env bash
+                echo INSIDE
+            "#,
+        )
+        .unwrap();
+
+    fs::File::create(&root.join("cargo-foo"))
+        .unwrap()
+        .write_all(
+            br#"!/usr/bin/env bash
+                echo OUTSIDE
+            "#,
+        )
+        .unwrap();
+
+    fs::set_permissions(root.join("cargo-foo"), fs::Permissions::from_mode(0o555)).unwrap();
+    fs::set_permissions(bin.join("cargo-foo"), fs::Permissions::from_mode(0o555)).unwrap();
+
+    cargo_process("foo")
+        .env("CARGO_HOME", &root)
+        .with_stdout_contains("INSIDE")
+        .run();
+
+    cargo_process("foo")
+        .env("PATH", format!("{}:{}", bin.display(), root.display()))
+        .env("CARGO_HOME", &root)
+        .with_stdout_contains("INSIDE")
+        .run();
+
+    println!("{}", root.display());
+    println!("{}", bin.display());
+    println!("{}", format!("{}:{}", root.display(), bin.display()));
+    cargo_process("foo")
+        .env("PATH", format!("{}:{}", root.display(), bin.display()))
+        .env("CARGO_HOME", &root)
+        .with_stdout_contains("OUTSIDE")
+        .run();
```

Comment:
Could you document the intention of each case?  Generally, I would do this by making them distinct tests and having the test name describe the intent

Reply:
I've split it in two, added some comments. Can split again or add more if you want.

Reply:
Personally I would prefer to keep it as a single test since there is a significant setup for this test.  There is a nontrivial overhead cost for each test, and this one in particular is building multiple executables which can take on the order of a second or two. With thousands of tests, that starts to add up to be very significant.

Reply:
In that case, I guess comments and/or code blocks to help clarify intent.

Really wish we had pytest's fixture system. since the test doesn't mutate the fixture...